End session when project of shared module is closed

tobous opened the pull request:
#### [FEATURE][CORE] Add logic to report projects as disposed

Adds the logic to report a project as disposed to the session. This
causes all activities for such projects to be dropped.

This functionality is useful to prevent the session from trying to apply
activities to disposed (closed) projects.

As this logic is currently only used by Saros/I, it is not set up to
handle disposed partial-sharing resources. If this would be come useful
at some point, the logic should be adjusted to also hold a mapping for
disposed partially-shared resources.

#### [INTERNAL][I] Explicitly catch and log EDT interruption exceptions

Adjusts IntelliJSynchronizer to catch and log ProcessCanceledExceptions
which are thrown when a synchronous execution on the EDT is interrupted.

#### [FIX][I] #582 End session when project of shared module is closed

Adds a listener to end the session when the project containing the
shared module is closed. The listener also reports the modules of the
project as disposed to the session to prevent it from trying to apply
any already queued activities for the disposed modules.

Fixes #582.

srossbach commented on core/src/saros/session/ISarosSession.java:
```diff
@@ -151,6 +151,16 @@ public void changePermission(User user, Permission permission)
+   *
+   * @param project the (disposed) shared project
+   */
+  default void disposeProject(IProject project) {
```
What happens if you just remove the project from the sharing ? I guess there will be inconsistency warnings and I think there is also no broadcast.

srossbach replied:
On the other hand as you are going to shutdown the session anyway, why do you not simply enable queuing ?

srossbach replied:
Ok nevermind, at least queuing will not work as it does not affect the queue in the ActivityHandler.

I would still prefer to change the name of this method. In general you do not want to dispose projects but initiate a half ordered shutdown (discarding pending activities). 

Do you have a good name proposal ?

tobous replied:
How about something along the lines of `notifyProjectDisposal` or `addActivityFilterForProject`?

As you have already mentioned, this does not have to mean that the session ends. Closing projects could could also be handled by removing the closed projects from the session.
But this should be handled by other parts of the implementation. This is only a workaround to stop the session from trying to apply activities to resources that are no longer available.

tobous replied:
Changed the wording to reflect the filter nature.

srossbach commented on intellij/src/saros/intellij/runtime/IntelliJSynchronizer.java:
```diff
@@ -32,7 +36,11 @@ private void exec(Runnable runnable, boolean async) {
+      try {
+        application.invokeAndWait(runnable, ModalityState.defaultModalityState());
+      } catch (ProcessCanceledException e) {
+        log.error("Synchronous execution on EDT interrupted - " + runnable, e);
```
What is interrupted ? The waiting for the execution to complete or the execution itself ?

tobous replied:
The [documentation of the `incokeAndWait(...)`](https://github.com/JetBrains/intellij-community/blob/master/platform/core-api/src/com/intellij/openapi/application/Application.java#L277) call says the following:
```Java
@throws ProcessCanceledException when the current thread is interrupted
```
So I...I guess both? Its not really specific...

srossbach commented on core/src/saros/session/internal/SarosSession.java:
```diff
@@ -164,6 +167,19 @@ public void send(List<User> recipients, IActivity activity) {
         @Override
         public void execute(IActivity activity) {
+          /*
+           * Filters out resource activities for disposed projects
```
What are disposed projects when translating this statement to Core context ?

tobous replied:
My bad. Missed that when removing the "disposed" references from the core stuff.

tobous commented on core/src/saros/session/internal/SarosSession.java:
```diff
@@ -99,6 +100,8 @@
   private final CopyOnWriteArrayList<IActivityProducer> activityProducers =
       new CopyOnWriteArrayList<IActivityProducer>();
 
+  private final Set<IProject> filteredProjects = new HashSet<>();
```
Done.

tobous commented on core/src/saros/session/ISarosSession.java:
```diff
@@ -151,6 +151,20 @@ public void changePermission(User user, Permission permission)
+   *
+   * @param project the shared project to add a filter for
+   */
+  default void addActivityFilterForProject(IProject project) {
```
Done. Changed to `setActivityExecution(IProject project, boolean enabled)`.

tobous commented:
Addressed review comments. And yay, I amended the wrong commit. Will fix this and push again.

Also, this PR currently does not address the teardown issues when an activity execution is already queued (throwing a `ProcessCanceledException). I still don't really know what to do about this. But as this is not bound to this PR, I would like to deal with it separately.

srossbach commented:
As I told you. Use a POISON pill. This should work after this patch.

tobous commented:
> As I told you. Use a POISON pill. This should work after this patch.

Yes, I got that (and the basic concept explained in the linked article, but I still don't quite know what exactly to do. But I will have a more thorough look at it and will get back to you if I still don't get it. :relaxed:

srossbach commented:
I am almost done, will push it in a few minutes.

tobous commented:
That works as well. Thanks. :relaxed:

tobous commented:
Moved fixes for review comments to correct commit.

srossbach commented on core/src/saros/session/ISarosSession.java:
```diff
@@ -151,6 +151,21 @@ public void changePermission(User user, Permission permission)
   public void removeListener(ISessionListener listener);
 
+  /**
+   * Enables or disables the activity execution for the given project. If the execution is disabled,
```
A bit misleading as one might assume it works for activities applied to this project and generated by this project. Should be fine though for the current issue.

tobous replied:
Adjusted to better reflect that this only applies to received activities.

srossbach commented on core/src/saros/session/internal/SarosSession.java:
```diff
@@ -164,6 +168,17 @@ public void send(List<User> recipients, IActivity activity) {
+          if (activity instanceof IResourceActivity) {
+            SPath path = ((IResourceActivity) activity).getPath();
+
+            if (path != null && filteredProjects.contains(path.getProject())) {
```
As far as I know there are EditorActivities with path = null to indicate an unknown editor.

tobous replied:
I only know of EditorActivities with `path=null` to indicate that no shared editor is selected/open.

But either way, if the path is `null`, we have no way of telling which project it belongs to, meaning we also can't apply it to a project. This would only be a problem if we have some activities that are applied to some "default project" such a case.

 But I will look through the `IResourceActivity` implementations to make sure.

tobous replied:
*Currently*, the only other resource activity besides `EditorActivity` that allows a `null` path is `JupiterActivity`, so it should not be a problem.

I am considering adding a comment to `AbstractResourceActivity` to avoid this being overlooked when adding new resource activities. But I am not sure whether it is really worth is.

tobous commented:
Rebased onto current master and amended javadoc for `setActivityExecution`.
